Enable metadata assertion and chart lock generation

komish opened the pull request:
- The owners module changes enables parsing Red Hat OWNERS files and asserting that metadata meets new expectations of the content.

- The packagemapping module provides hooks to generating chart locks based on the content of a given repository's charts/ directory.

- The pullrequest module changes facilitates asserting metadata given a pull request as an input.

- Entrypoints are then added to setup.cfg to simplify calling of this new logic in CI.

Depends on https://github.com/openshift-helm-charts/development/pull/313 for Python Style to pass.

mgoerens commented on scripts/src/packagemapping/generatelocks.py:
```diff
@@ -0,0 +1,106 @@
+            sys.exit(1)
+
+        with open(filename, "r") as f:
+            d = yaml.load(f.read(), Loader=yaml.BaseLoader)
+            owners_value_chart_name = d["chart"]["name"]
+            owners_value_vendor_label = d["vendor"]["label"]
```
Could / Should we reuse the functions available in the `owners` package for getting the name and vendor label for this chart ?

Otherwise, we probably want to catch the Exceptions in case e.g. the Yaml is not correctly formatted (`yaml.load` fails) or some key is not present in the file, so we can set a proper exit code and error message.

komish replied:
+1, will fix this in the next push.

mgoerens commented on scripts/src/packagemapping/generatelocks.py:
```diff
@@ -0,0 +1,106 @@
+            )
+
+        new_entry = f"{category}/{organization}/{chart}"
+        if packages.get(chart, None) is not None:
```
```suggestion
        if packages.get(chart) is not None:
```

This is enough as `None` is the default return value if the provided key cannot be found in the dict.

mgoerens commented on scripts/src/pullrequest/metadata.py:
```diff
@@ -0,0 +1,150 @@
+            return 0
+        print(f"[Error] {e}", file=sys.stderr)
+        return 20
+    except Exception as e:
+        print("[Error] Something unexpected went wrong", e, file=sys.stderr)
+        return 90
```
Same comment as in generatelocks.py

mgoerens commented on scripts/src/pullrequest/metadata.py:
```diff
@@ -0,0 +1,150 @@
+        )
+
+
+class MultipleMatchesError(Exception):
```
We already have checks in place (e.g. [here](https://github.com/komish/charts-development/blob/3f3fa8485b6d5e2a11641b89a6d1e757620f732f/scripts/src/checkprcontent/checkpr.py#L159) ) to ensure that a PR relates to only one chart.

Is this needed here ? Could we find a way to put the code in common ? This kind of relates to #296

komish replied:
This library definitely has common code, but focuses just on the extraction of category, org, chart, etc. This is just one component of the linked code ( checkprcontent ), which also checks tarball naming, if relevant.

I do think this library can be leveraged there, but the additional checks need to be accounted for in some way, and I'm not ready to introduce changes to that code path if it's not strictly necessary. At least not as a part of this larger bugfix effort.

I'd vote for pursuing this in #296 as a future action item.

mgoerens reviewed (changes requested):
Functionally, no comment :+1: 

Some changes are needed IMO in the code, for example when catching `Exception` in main + suggested other minor changes.

komish commented on scripts/src/owners/redhat_metadata.py:
```diff
@@ -0,0 +1,89 @@
+
+
+def main():
+    """The CLI entrypoint. Returns an exit code."""
```
Will do, but I would argue that I don't consider them overly important. Incrementing exit codes in this (or a similar) way is an old habit, for me, originating in the time when some programs would throw nondescript error messages along with very meaningful error codes.

I generally avoid making the exit codes part of the API (by strongly defining them in documentation), but still tend to increment them because "some developer" may go back to the source and see "ah, it exited 20 which means this".

All that to say - yes, I'll happily document it here, but it's an old habit that doesn't have much backing considering I'm returning meaningful (I hope) error messages :laughing:

komish commented on scripts/src/packagemapping/generatelocks.py:
```diff
@@ -0,0 +1,106 @@
+from glob import glob
+import sys
+import yaml
+import re
+from datetime import datetime, timezone
+from json import dumps as to_json
```
:+1: Understood.

This should be tool-enforced. 

As such, for this PR,  I'll fix this using an extra linter in `ruff` for the time being, and then circle back and implement the same linter into the base ruff.toml along with application of the same rule across the entire scripts/src/ directory.

komish replied:
The linter integration PR is here: https://github.com/openshift-helm-charts/development/pull/315

komish commented on scripts/src/owners/redhat_metadata.py:
```diff
@@ -0,0 +1,89 @@
+def assert_redhat_metadata(contents):
+    """Ensures contents of owners file for Red Hat submissions
+
+    Red Hat submissions are expected to have certain metadata. - The owner label
+    must be "redhat" - The owner organizaton name must be "Red Hat" - The
+    chart's name must be prefixed with "redhat-"
```
Yes! This will be addressed in the next push.

komish commented on scripts/src/owners/redhat_metadata.py:
```diff
@@ -0,0 +1,89 @@
+        _, ownerdata = owners_file.get_owner_data(
+            args.category, args.organization, args.chartname
+        )
+    except Exception as e:
+        print(e)
+        return 10
```
Great point. This will be addressed in the next push.

mgoerens replied:
Thanks, also opened https://github.com/openshift-helm-charts/development/issues/317 as IMO the more pythonic way to do this would be to raise a custom Exception rather than set a boolean to False if there was an error. To be discussed (nothing major of course)

komish commented on scripts/src/owners/redhat_metadata.py:
```diff
@@ -0,0 +1,89 @@
+    except RedHatOwnersFileInvalidContentsError as e:
+        print(f"[Error] Invalid contents, {e}")
+        return 20
+    except Exception as e:
+        print(f"[Error] Unexpected error: {e}")
+        return 90
```
Generally, I have felt that having a fallback/default exception handler is a must. The called functions may not have exceptions outside of those accounted for already, but should that change, it's possible a developer may modified a called function such that it returns an exception not already accounted for. If it isn't accounted for, the following behavior will be unexpected.

As such, I've typically included the fallback exception block.

With that said, what you say is reasonable, and what I say could be perceived as a premature optimization. This exception handler will be removed in the next push.

komish commented on scripts/src/packagemapping/generatelocks.py:
```diff
@@ -0,0 +1,106 @@
+    return pattern
+
+
+def logError(msg, file=sys.stderr):
+    """logError just prints the msg with an ERROR caption to stderr unless otherwise defined"""
+    print(f"[ERROR] {msg}", file=file)
+
+
+def logInfo(msg, file=sys.stderr):
+    """logError just prints the msg with an INFO caption to stderr unless otherwise defined"""
+    print(f"[INFO] {msg}", file=file)
+
+
+def logWarn(msg, file=sys.stderr):
+    """logWarn just prints the msg with an INFO caption to stderr unless otherwise defined"""
+    print(f"[WARN] {msg}", file=file)
```
I 100% agree. These were added in the early days of this hackery.

I don't want to go through and refactor everything to use these now, so if you'll allow it, I've created https://github.com/openshift-helm-charts/development/issues/316 to tackle this at a later date.

mgoerens replied:
_Insert Community gif: "I'll allow it"_

komish commented on scripts/src/packagemapping/generatelocks.py:
```diff
@@ -0,0 +1,106 @@
+                f"did not successfully parse the input. Did you run this in the right place? filename: {filename}",
+            )
+            # continue
+            sys.exit(1)
```
Great catch. This is one of the earliest things I wrote in this effort, and as such, as potentially the messiest. I'll resolve this in the next push.

komish commented on scripts/src/packagemapping/generatelocks.py:
```diff
@@ -0,0 +1,106 @@
+        pattern = ownerfile_regex()
+        matched = pattern.match(filename)
+        if matched is None:
+            # TODO(komish): Consider if we need to have a flag that exits and otherwise make this a warning
```
Definitely -

When this script was first written, it had to combat multiple chart directories that had invalid names. There were no charts associated with these, presumably, because of the incorrect naming scheme. A good example is a chart that has a name starting with a bracket, e.g. `[`. This is invalid in multiple places; I'm pretty sure a chart cannot start with the `[` character.

The original intention here was to allow for those to exist in repo but still not cause workflows that ran this script to fail, such that a lock would not be accounted for when parsing the failing entry.

In the grand scheme, however, I think this comment can and should be removed. We pretty much always want this to fail in cases where the input is invalid. I'll remove these lines in the next push.

komish commented on scripts/src/owners/owners_file.py:
```diff
@@ -33,6 +33,15 @@ def get_vendor(owner_data):
+    vendor = ""
+    try:
+        vendor = owner_data["vendor"]["label"]
+    except Exception:
```
I think this can be rewritten to be more functional for us as a whole.

With that said, I want to maintain this change to be as minimal as possible for the purposes of reducing modified code paths. TL;DR I'd like to improve this in the future, unrelated to this bugfix.

mgoerens replied:
Opened https://github.com/openshift-helm-charts/development/issues/317 to track this future work.

komish commented:
@mgoerens I believe I've addressed all of the issues that I've agreed should be addressed here.
I've left things as individual commits here because we're making a lot of changes, and I wanted to make it a bit easier to track what change was motivated by which code review comment.

mgoerens commented on scripts/src/packagemapping/generatelocks.py:
```diff
@@ -0,0 +1,121 @@
+        owners_value_chart_name = owners_file.get_chart(ownersContent)
+        owners_value_vendor_label = owners_file.get_vendor_label(ownersContent)
+
+        if owners_value_chart_name != chart:
+            logError(
+                f"the chart name in the OWNERS file did not match the chart name directory structure. OWNERS_FILE_VALUE={owners_value_chart_name}, DIRECTORY_VALUE:{chart}"
+            )
```
On top of logging an Error, do we need to stop here ? Same with the next block, that checks if the vendor label matches the organization name directory structure.

komish replied:
We can. I'm trying to think; this will most likely impact Red Hat charts because their chart names will have a different requirement moving forwrad (the prefix). I would need to correct their structure in order to not fail here.

Let's go ahead and throw an error for now. I'll push this fix shortly.

komish replied:
This change has been pushed.

mgoerens commented on scripts/src/pullrequest/metadata.py:
```diff
@@ -0,0 +1,155 @@
+        dest="emit_to_github",
+        action="store_true",
+    )
+    parser.add_argument(
+        "--ignore-error-no-matches",
+        help="When enabled, returns a zero exit code instead of a non-zero for cases where no matches were found",
+        dest="ignore_no_matches",
+        action="store_true",
+        default=False,
+    )
```
Is this flag motivated by an actual use case ?

komish replied:
I don't see a use case in it in my dev repo, and I don't remember why I included it off hand. If you'd prefer it, I can remove the flag.

mgoerens replied:
Yes, I would be in favor of removing this code if it's not needed.

komish replied:
Fixed and pushed here: https://github.com/openshift-helm-charts/development/pull/314/commits/035b0274a32185316345cb8b7926bac5f35d35cd

Tests have been canceled and should be rerunning now.

mgoerens reviewed (commented):
All comments were addressed :+1: 

I left a couple of new questions, regarding things I didn't think of during the first review.
